src/compiler/checker.ts

```diff
@@ -11395,7 +11395,12 @@ namespace ts {
                         // such that direct inferences to T get priority over inferences to Partial<T>, for example.
                         const inference = getInferenceInfoForType((<IndexType>constraintType).type);
                         if (inference && !inference.isFixed) {
-                            const inferredType = inferTypeForHomomorphicMappedType(source, <MappedType>target);
+                            if (contains(mappedTypeStack, [source, target.symbol], ([s1, t1], [s2, t2]) => s1 === s2 && t1 === t2)) {
+                                return;
+                            }
+                            (mappedTypeStack || (mappedTypeStack = [])).push([source, target.symbol]);

```

Comment:
```ts
function getInferenceStackKey(source: Type, target: Symbol) {
    return `${source.id},${getSymbolId(target)}`
}
// ...
const cacheKey = getInferenceStackKey(source, target.symbol);
if (mappedTypeCache && mappedTypeCache.has(cacheKey)) {
    return; // future: mappedTypeCache.get(cacheKey)
}
(mappedTypeCache || (mappedTypeCache = createMap())).set(cacheKey, true); // future: replace `true` with initialized candidate type which is filled out by the below `infer` call.
const inferredType = inferTypeForHomomorphicMappedType(source, <MappedType>target, mappedTypeCache);
mappedTypeCache.delete(cacheKey);
```
If the `mappedTypeStack` values are never needed other than for set membership checking, and the ordering is irrelevant other than for removal, then for efficiency (faster contains checks, no array allocation for each member) we should use a set and not a stack, right?

Reply:
Yes. I was going to do that next, but I decided to get @ahejlsberg's opinion about the general approach beforehand.